### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by canthefason**

### CODE REVIEW

**Comment by cihangir** [comment]
Add travis badge

File: accessgrant.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+		New: func() interface{} {
+			return NewPubNubClient(cs)
+		},
+	}
+
+	return &AccessGrant{pool: p}
+}
+
+func (ag *AccessGrant) Grant(as *AuthSettings) error {
+	client := ag.pool.Get().(*PubNubClient)
```
Comment (cihangir) [inline accessgrant.go]: from the sync.Pool doc `Callers should not assume any relation between values passed to Put and the values returned by Get.` namely better to check if the interface conversion is `ok`

File: accessgrant_test.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package pubnub
+
+import "testing"
+
+func TestGrantAccessWithGranter(t *testing.T) {
+	cs := new(ClientSettings)
+	cs.ID = "-1"
+	cs.SubscribeKey = "subscribekey"
```
Comment (cihangir) [inline accessgrant_test.go]: you can read those from env variables

File: channel.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	go p.pub.Subscribe(name, "", pr.successCh, false, pr.errorCh)
+
+	if err := pr.Do(); err != nil {
+		return nil, err
+	}
+
+	return channel, nil
+}
+
+func (c *Channel) Close() {
```
Comment (cihangir) [inline channel.go]: are we gonna do smh else with channel? just open/close?

File: client.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		return channel, nil
+	}
+
+	channel, err := p.NewChannel(channelName)
+	if err != nil {
+		return nil, err
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
```
Comment (cihangir) [inline client.go]: instead of deferring you could put unlock under writing

File: client.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	p.channels[channelName] = channel
+
+	return channel, nil
+}
+
+//////////////////// PubnubAuthSettings /////////////////////
+
+type AuthSettings struct {
+	// PubNub channel name
```
Comment (cihangir) [inline client.go]: as godoc convention, there should be new lines between each properties

File: channel.go (hunk 3 above)
Comment (canthefason) [inline channel.go]: actually it was used for consuming channels messages, but for this purpose adding a Consume method makes more sense

File: client_test.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	receiver.Subscribe("sueme")
+
+	channel, err := receiver.Subscribe("testme")
+	if err != nil {
+		t.Errorf("Expected nil but got error while subscribing: %s", err)
+		t.FailNow()
+	}
+
+	var wg sync.WaitGroup
+	go func() {
```
Comment (cihangir) [inline client_test.go]: we are gonna add more tests here

File: request.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+		// this is an http response
+		pr.parseGrantResponse(r.(map[string]interface{}))
+	}
+
+}
+
+func (pr *PubNubRequest) parseGrantResponse(response map[string]interface{}) {
+	isError, ok := response["error"].(bool)
+	if !ok {
+		isError = false
```
Comment (cihangir) [inline request.go]: default is false:P

File: client.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+func (p *PubNubClient) Subscribe(channelName string) (*Channel, error) {
+	if channelName == "" {
+		return nil, ErrChannelNotSet
+	}
+
+	return p.fetchOrCreateChannel(channelName)
+}
+
+func (p *PubNubClient) Close() {
+	p.pub.CloseExistingConnection()
```
Comment (cihangir) [inline client.go]: after calling `Close()` no other operations should be allowed

**Comment by canthefason** [comment]
except the travis badge, PR is updated. I will do that some other time.

**Pull request merged**